pkg/package-server/provider/registry.go

```diff
@@ -10,108 +10,135 @@ import (
 )
 
 const (
-	defaultConnectionTimeout = 5 * time.Second
+	catalogIndex = "catalog"
+	cacheTimeout = 5 * time.Minute
+	readyTimeout = 10 * time.Minute
+	stateTimeout = 20 * time.Second
 )
 
-type sourceKey struct {
-	name      string
-	namespace string
+func getSourceKey(pkg *operators.PackageManifest) (key *resolver.CatalogKey) {
+	if pkg != nil {
+		key = &resolver.CatalogKey{
+			Namespace: pkg.Status.CatalogSourceNamespace,
+			Name:      pkg.Status.CatalogSource,
+		}
+	}
+
+	return
+}
+
+func catalogIndexFunc(obj interface{}) ([]string, error) {
+	pkg, ok := obj.(*operators.PackageManifest)
+	if !ok {
+		return []string{""}, fmt.Errorf("obj is not a packagemanifest %v", obj)
+	}
+
+	return []string{getSourceKey(pkg).String()}, nil
 }
 
 type registryClient struct {
 	api.RegistryClient
-	source *operatorsv1alpha1.CatalogSource
+	catsrc *operatorsv1alpha1.CatalogSource
 	conn   *grpc.ClientConn
 }
 
-func newRegistryClient(source *operatorsv1alpha1.CatalogSource, conn *grpc.ClientConn) registryClient {
-	return registryClient{
+func newRegistryClient(catsrc *operatorsv1alpha1.CatalogSource, conn *grpc.ClientConn) *registryClient {
+	return &registryClient{
 		RegistryClient: api.NewRegistryClient(conn),
-		source:         source,
+		catsrc:         catsrc,
 		conn:           conn,
 	}
 }
 
+func (r *registryClient) key() (key resolver.CatalogKey, err error) {
+	if r.catsrc == nil {
+		err = fmt.Errorf("cannot get key, nil catalog")
+		return
+	}
+
+	key = resolver.CatalogKey{
+		Namespace: r.catsrc.GetNamespace(),
+		Name:      r.catsrc.GetName(),
+	}
+
+	return
+}
+
 // RegistryProvider aggregates several `CatalogSources` and establishes gRPC connections to their registry servers.
 type RegistryProvider struct {
 	queueinformer.Operator
+	runOnce sync.Once
 
-	mu              sync.RWMutex
 	globalNamespace string
-	clients         map[sourceKey]registryClient
+	sources         *registrygrpc.SourceStore
+	cache           cache.Indexer
+	pkgLister       pkglisters.PackageManifestLister
+	catsrcLister    operatorslisters.CatalogSourceLister
 }
 
 var _ PackageManifestProvider = &RegistryProvider{}
 
-func NewRegistryProvider(ctx context.Context, crClient versioned.Interface, operator queueinformer.Operator, wakeupInterval time.Duration, watchedNamespaces []string, globalNamespace string) (*RegistryProvider, error) {
+func NewRegistryProvider(ctx context.Context, crClient versioned.Interface, operator queueinformer.Operator, wakeupInterval time.Duration, globalNamespace string) (*RegistryProvider, error) {
 	p := &RegistryProvider{
 		Operator: operator,
 
 		globalNamespace: globalNamespace,
-		clients:         make(map[sourceKey]registryClient),
+		cache: cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{
+			cache.NamespaceIndex: cache.MetaNamespaceIndexFunc,
+			catalogIndex:         catalogIndexFunc,
+		}),
 	}
-
-	for _, namespace := range watchedNamespaces {
```

Comment:
I got rid of multi-namespace configuration, since other OLM controllers already assume cluster-wide privileges. I also did this to avoid needing to register/access different `CatalogSource` listers for each namespace.